### PR TITLE
update the backward compatibility guide

### PR DESCRIPTION
**Pull request by fascinosum**
## This PR is a:

- [ ] New topic
- [x] Content update
- [ ] Content fix or rewrite
- [ ] Bug fix or improvement

## Summary

When this pull request is merged, it will...

## Additional information

List all affected URLs 

- https://devdocs.magento.com/guides/v2.1/contributor-guide/backward-compatible-development/

### CODE REVIEW

**Comment by magento-cicd2** [comment]
An admin must run tests on this PR before it can be merged.

File: guides/v2.1/contributor-guide/backward-compatible-development/index.md [hunk 1]
```diff
@@ -39,6 +39,12 @@ Mark the method with the `@deprecated` tag instead of removing it.
 
 Continue returning the same results from the method if possible, so the old functionality is preserved.
 
+If you need to remove the existing functionality and you do not need alternative implementation, you SHOULD change the method code to
+
+```php
+trigger_error('Function is deprecated', E_USER_WARNING);
```
Comment (buskamuza) [inline guides/v2.1/contributor-guide/backward-compatible-development/index.md]: Deprecation is not what happens in such situation, so I think deprecation warning will be misleading.
I'd recommend real exception or, even better, removal of entire method in such situation. So that it's visible to everyone that there is a breaking change.

File: guides/v2.1/contributor-guide/backward-compatible-development/index.md [hunk 2]
```diff
@@ -39,6 +39,12 @@ Mark the method with the `@deprecated` tag instead of removing it.
 
 Continue returning the same results from the method if possible, so the old functionality is preserved.
 
+If you need to remove the existing functionality and you do not need alternative implementation, you SHOULD change the method code to
```
Comment (buskamuza) [inline guides/v2.1/contributor-guide/backward-compatible-development/index.md]: I'd rephrase, "and you do not need alternative implementation" to "and there is no alternative implementation" because developer of this part of code obviously don't need this functionality, but we need to care about other developers, and they may need it. So usually we should preserve the functionality even if core Magento doesn't need it. The only exception is when it's impossible to provide an alternative implementation.

Comment (buskamuza) [inline guides/v2.1/contributor-guide/backward-compatible-development/index.md]: Also, "SHOULD" should be changed to "MUST". There are no exceptions to this rule.

**Comment by magento-cicd2** [comment]
An admin must run tests on this PR before it can be merged.

**Comment by hguthrie** [comment]
@fascinosum any status on this?

File: guides/v2.1/contributor-guide/backward-compatible-development/index.md (hunk 1 above)
Comment (fascinosum) [reply guides/v2.1/contributor-guide/backward-compatible-development/index.md]: `trigger_error` is much better than Exception, you can't catch the user error in the catch block. Removing a  method is a backward incompatible implementation and sometimes it can't be implemented, when you have an inheritance

File: guides/v2.1/contributor-guide/backward-compatible-development/index.md [hunk 3]
```diff
@@ -39,6 +39,12 @@ Mark the method with the `@deprecated` tag instead of removing it.
 
 Continue returning the same results from the method if possible, so the old functionality is preserved.
 
+If you need to remove the existing functionality and there is no alternative implementation, you MUST change the method code to
```
Comment (joni-jones) [inline guides/v2.1/contributor-guide/backward-compatible-development/index.md]: Actually, you can ADD `trigger_error` method but don't replace the existing. The method should still return the same result.

Comment (joni-jones) [inline guides/v2.1/contributor-guide/backward-compatible-development/index.md]: Actually, you can **add** `trigger_error` method but don't replace the existing. The method should still return the same result.

**Comment by magento-cicd2** [comment]
An admin must run tests on this PR before it can be merged.

File: guides/v2.1/contributor-guide/backward-compatible-development/index.md (hunk 3 above)
Comment (fascinosum) [reply guides/v2.1/contributor-guide/backward-compatible-development/index.md]: It is impossible to return the same result and trigger error at the same time. I've changed the type of error

Comment (joni-jones) [reply guides/v2.1/contributor-guide/backward-compatible-development/index.md]: What type of error are you trying to use? Doesn't it work with `E_WARNING`?

**Comment by magento-cicd2** [comment]
An admin must run tests on this PR before it can be merged.

File: guides/v2.1/contributor-guide/backward-compatible-development/index.md (hunk 3 above)
Comment (fascinosum) [reply guides/v2.1/contributor-guide/backward-compatible-development/index.md]: `E_USER_ERROR` is a real PHP error and this avoids hidden warnings in the log

Comment (joni-jones) [reply guides/v2.1/contributor-guide/backward-compatible-development/index.md]: @fascinosum why to don't use `E_DEPRECATED`? PHP core team uses it to "mark" deprecated functionality. Why the deprecated code should throw an error?

Comment (fascinosum) [reply guides/v2.1/contributor-guide/backward-compatible-development/index.md]: @joni-jones because this is not a deprecated code, but a backward incompatible change. You must not use this method anymore

Comment (joni-jones) [reply guides/v2.1/contributor-guide/backward-compatible-development/index.md]: BC guide should not contain approaches which introduce BIC.

Comment (akaplya) [reply guides/v2.1/contributor-guide/backward-compatible-development/index.md]: Why should we keep a method if we can not mimic an execution result?
